core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,6 +73,51 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
       }
     }
 
+  final def foldM[R1 <: R, E2, B](
+    failure: E => ZManaged[R1, E2, B],
+    success: A => ZManaged[R1, E2, B]
+  ): ZManaged[R1, E2, B] =
+    ZManaged[R1, E2, B] {
+      Ref.make[ZIO[R1, Nothing, Any]](IO.unit).map { finalizers =>
+        Reservation(
+          acquire = {
+            val direct =
+              ZIO.uninterruptibleMask { k =>
+                k(self.reserve)
+                  .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                  .flatMap(res => k(res.acquire))
+              }
+            val onFailure = (e: E) =>
+              ZIO.uninterruptibleMask { k =>
+                k(failure(e).reserve)
+                  .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                  .flatMap(res => k(res.acquire))
+              }
+            val onSuccess = (a: A) =>
+              ZIO.uninterruptibleMask { k =>
+                k(success(a).reserve)
+                  .flatMap(res => finalizers.update(fs => res.release *> fs).const(res))
+                  .flatMap(res => k(res.acquire))
+              }
+            direct.foldM(onFailure, onSuccess)
+          },
+          release = ZIO.flatten(finalizers.get)
+        )
+      }
+    }
+
+  final def catchAll[R1 <: R, E2, A1 >: A](h: E => ZManaged[R1, E2, A1]) =
```

Comment:
Love to see these methods on `ZManaged`! Over time, `ZManaged` should become about as powerful as `ZIO`...

Reply:
Agreed. I plan to do another pr shortly after where I just port all the stuff from ZIO to ZManaged